src/codegen/baseline_jit.h

```diff
@@ -0,0 +1,249 @@
+// it can get executed by a jump from another fragment.
+// At every fragment end we can jump to another fragment, fallback to the Interpreter or exit.
+// This means we are not allowed to assume that a register contains a specific value between JitFragments.
+// This also means that we are allowed to store a Python variable which only lives in the current CFGBLock* inside a
+// register or stack slot but we aren't if it outlives the block - we have to store it in the interpreter instance.
+//
+// To execute a specific CFGBlock one has to call:
+//      CFGBlock* block;
+//      block->entry_code(ast_interpreter_instance, block)
+//
+// Signature of a JitCodeBlock:
+//  std::pair<CFGBlock*, Box*>(*entry_code)(ASTInterpreter* interp, CFGBlock* block)
+//  args:
+//      interp:  instance to the ASTInterpreter
+//      block:   block to execute
+//
+//  return value:
+//      first:   next block to execute in the interpreter
+//                  if 0, we executed a return node and should return second
+//      second:  return value only used if first == 0
+//
+// Basic layout of generated code block is:
+// entry_code:
+//      push   %rbp                 ; setup frame pointer
+//      mov    %rsp,%rbp
+//      sub    $0x108,%rsp          ; setup scratch, 0x108 = scratch_size + 8 (=stack alignment)
+//      push   %rdi                 ; save the pointer to ASTInterpreter instance
+//      jmpq   *0x8(%rsi)           ; jump to block->code
+//                                      possible values: first_JitFragment, second_JitFragment,...
+//
+// first_JitFragment:
+//      ...
+//    ; Side exit (e.g. if <val> <block1> else <block2>
+//      movabs $0x1270014108,%rcx   ; rcx = True
+//      cmp    %rax,%rcx            ; rax == True
+//      jne    end_side_exit
+//      mov    %rax,0x10(%rsp)      ;
+//      movabs $0x215bb60,%rax      ; rax = CFGBlock* to call next (rax is also the 1. return reg)
+//      mov    0x8(%rax),%rsi       ; load CFGBlock->code
+//      test   %rsi,%rsi            ; CFGBlock->code == 0
+//      je     epilog               ; exit to interpreter if code == 0
+//      jmpq   *0x8(%rax)           ; jump to new jit fragment (e.g second_JitFragment)
+//    end_side_exit:
+//      ....
+
+// second_JitFragment:
+//      ...
+//
+// nth_JitFragment:
+//      ...                         ; direct jump previous JITed block
+//      jmp first_JitFragment
+//
+// epilog:                          ; code which jumps to epilog has to make sure that
+//                                  ; rax contains the next block to execute
+//                                  : or 0 if we are finished but then rdx must contain the Box* value to return
+//      leave
+//      ret
+//
+class JitCodeBlock {
+private:
+    static constexpr int scratch_size = 256;
+    static constexpr int code_size = 4096 * 2;
+    static constexpr int epilog_size = 2; // size of [leave, ret] in bytes
+
+    void* code;
```

Comment:
Maybe this could be a unique_ptr?  I know we're not freeing any of the memory yet, but delineating the ownership should still be a good thing.